File: src/main/java/waterpower/integration/jei/JeiModule.java
package waterpower.integration.jei;

import mezz.jei.api.BlankModPlugin;
import mezz.jei.api.IGuiHelper;
import mezz.jei.api.IJeiRuntime;
import mezz.jei.api.IModPlugin;
import mezz.jei.api.IModRegistry;
import mezz.jei.api.ISubtypeRegistry;
import mezz.jei.api.JEIPlugin;
import mezz.jei.api.ingredients.IModIngredientRegistration;
import mezz.jei.api.recipe.IRecipeCategory;
import mezz.jei.api.recipe.IRecipeHandler;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;
import waterpower.client.Local;
import waterpower.common.block.GlobalBlocks;
import waterpower.common.block.machines.GuiCompressor;
import waterpower.common.block.machines.GuiCutter;
import waterpower.common.block.machines.GuiLathe;
import waterpower.common.block.machines.GuiMacerator;
import waterpower.common.block.machines.GuiSawmill;
import waterpower.common.recipe.IRecipeManager;
import waterpower.common.recipe.MyRecipes;

@JEIPlugin
public class JeiModule extends BlankModPlugin {

	@Override
	public void register(IModRegistry registry) {
		IGuiHelper guiHelper = registry.getJeiHelpers().getGuiHelper();
		registry.addRecipeHandlers(new MyRecipeHandler());
		addMachineRecipes(registry, "waterpower.compressor", GuiCompressor.TITLE_I18N, MyRecipes.compressor, guiHelper, GuiCompressor.GUI, GlobalBlocks.compressor, GuiCompressor.class);
		addMachineRecipes(registry, "waterpower.cutter",GuiCutter.TITLE_I18N,  MyRecipes.cutter, guiHelper, GuiCutter.GUI, GlobalBlocks.cutter, GuiCutter.class);
		addMachineRecipes(registry, "waterpower.lathe", GuiLathe.TITLE_I18N, MyRecipes.lathe, guiHelper, GuiLathe.GUI, GlobalBlocks.lathe, GuiLathe.class);
		addMachineRecipes(registry, "waterpower.macerator", GuiMacerator.TITLE_I18N, MyRecipes.macerator, guiHelper, GuiMacerator.GUI, GlobalBlocks.macerator, GuiMacerator.class);
		addMachineRecipes(registry, "waterpower.sawmill", GuiSawmill.TITLE_I18N, MyRecipes.sawmill, guiHelper, GuiSawmill.GUI, GlobalBlocks.sawmill, GuiSawmill.class);
	}

	private void addMachineRecipes(IModRegistry registry, String uid, String titleI18N, IRecipeManager recipeManager, IGuiHelper guiHelper, ResourceLocation image, ItemStack itemStack, Class guiClass) {
		MyRecipeCategory category = new DefaultMyRecipeCategory(uid, Local.get(titleI18N), recipeManager, guiHelper, image);
		registry.addRecipeCategories(category);
		registry.addRecipes(MyRecipeWrapper.createRecipes(category));
		registry.addRecipeClickArea(guiClass, 50, 34, 24, 14, uid);
		registry.addRecipeCategoryCraftingItem(itemStack, uid);
	}

}

File: src/main/kotlin/waterpower/integration/jei/JEIModule.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.integration.jei

import mezz.jei.api.BlankModPlugin
import mezz.jei.api.IGuiHelper
import mezz.jei.api.IModRegistry
import mezz.jei.api.JEIPlugin
import net.minecraft.client.gui.inventory.GuiContainer
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import waterpower.WaterPower
import waterpower.client.i18n
import waterpower.common.block.machine.*
import waterpower.common.init.WPBlocks
import waterpower.common.recipe.IRecipeManager
import waterpower.common.recipe.RecipeManagers

@JEIPlugin
class JEIModule : BlankModPlugin() {

    override fun register(registry: IModRegistry) {
        val guiHelper = registry.jeiHelpers.guiHelper
        registry.addRecipeHandlers(RecipeHandler())
        addMachineRecipes(registry, "advcompressor", RecipeManagers.advCompressor, guiHelper, WPBlocks.advanced_compressor, GuiAdvCompressor::class.java)
        addMachineRecipes(registry, "compressor", RecipeManagers.compressor, guiHelper, WPBlocks.compressor, GuiCompressor::class.java)
        addMachineRecipes(registry, "cutter", RecipeManagers.cutter, guiHelper, WPBlocks.cutter, GuiCutter::class.java)
        addMachineRecipes(registry, "lathe", RecipeManagers.lathe, guiHelper, WPBlocks.lathe, GuiLathe::class.java)
        addMachineRecipes(registry, "crusher", RecipeManagers.crusher, guiHelper, WPBlocks.crusher, GuiCrusher::class.java)
        addMachineRecipes(registry, "sawmill", RecipeManagers.sawmill, guiHelper, WPBlocks.sawmill, GuiSawmill::class.java)
    }

    private fun addMachineRecipes(registry: IModRegistry, id: String, recipeManager: IRecipeManager, guiHelper: IGuiHelper, itemStack: ItemStack, guiClass: Class<out GuiContainer>) {
        val uid = "waterpower.machine.$id"
        val category = DefaultRecipeCategory(uid, i18n(uid), recipeManager, guiHelper, ResourceLocation("${WaterPower.MOD_ID}:textures/gui/$id.png"))
        registry.addRecipeCategories(category)
        registry.addRecipes(RecipeWrapper.createRecipes(category))
        registry.addRecipeClickArea(guiClass, 50, 34, 24, 14, uid)
        registry.addRecipeCategoryCraftingItem(itemStack, uid)
    }

}
